web/src/message_store.ts

```diff
@@ -0,0 +1,277 @@
+export type MessageWithBooleans = (
+    | Omit<RawMessage & {type: "private"}, "flags">
+    | Omit<RawMessage & {type: "stream"}, "flags">
+) & {
+    unread: boolean;
+    historical: boolean;
+    starred: boolean;
+    mentioned: boolean;
+    mentioned_me_directly: boolean;
+    stream_wildcard_mentioned: boolean;
+    topic_wildcard_mentioned: boolean;
+    collapsed: boolean;
+    alerted: boolean;
+};
+
+export type MessageCleanReaction = {
+    class: string;
+    count: number;
+    emoji_alt_code: boolean;
+    emoji_code: string;
+    emoji_name: string;
+    is_realm_emoji: boolean;
+    label: string;
+    local_id: string;
+    reaction_type: string;
+    user_ids: number[];
+    vote_text: string;
+};
+
+export type Message = (
+    | Omit<MessageWithBooleans & {type: "private"}, "reactions">
+    | Omit<MessageWithBooleans & {type: "stream"}, "reactions">
+) & {
+    // Added in `reactions.set_clean_reactions`.
+    clean_reactions: Map<string, MessageCleanReaction>;
+
+    // Added in `message_helper.process_new_message`.
+    sent_by_me: boolean;
+    reply_to: string;
+    display_reply_to?: string;
+
+    // These properties are used in `message_list_view.js`.
+    starred_status: string;
+    message_reactions: MessageCleanReaction[];
+    url: string;
+
+    // Used in `markdown.js`, `server_events.js`, and `set_message_booleans`
+    flags?: string[];
+} & (
+        | {
+              type: "private";
+              is_private: true;
+              is_stream: false;
+              pm_with_url: string;
+              to_user_ids: string;
+          }
+        | {
+              type: "stream";
+              is_private: false;
+              is_stream: true;
+              stream: string;
+              topic: string;
+          }
+    );
+
+export function update_message_cache(message: Message): void {
+    // You should only call this from message_helper (or in tests).
+    stored_messages.set(message.id, message);
+}
+
+export function get_cached_message(message_id: number): Message {
+    // You should only call this from message_helper.
+    // Use the get() wrapper below for most other use cases.
+    return stored_messages.get(message_id);
+}
+
+export function clear_for_testing(): void {
+    stored_messages.clear();
+}
+
+export function get(message_id: number): Message | undefined {
+    if (message_id === undefined || message_id === null) {
+        blueslip.error("message_store.get got bad value", {message_id});
+        return undefined;
+    }
+
+    if (typeof message_id !== "number") {
+        blueslip.error("message_store got non-number", {message_id});
+
+        // Try to soldier on, assuming the caller treats message
+        // ids as strings.
+        message_id = Number.parseFloat(message_id);
+    }
+
+    return stored_messages.get(message_id);
+}
+
+export function get_pm_emails(message: Message): string {
+    const user_ids = people.pm_with_user_ids(message) || [];
```

Comment:
In modern JS, when targeting `undefined`, we should always prefer nullish coalescing `??` over a logical operator `||`.

(`??` specifically targets `null` and `undefined`, while `||` targets `null`, `undefined`, `false`, `NaN`, `0`, `-0`, `0n`, `""`, and `document.all`. The latter leads to bugs when dealing with a `number | undefined` that might be `0` or a `string | undefined` that might be `""`, so it’s better to prefer the safer pattern for all types.)

Reply:
Oh yes, I've been using this places but forgot this time. Thanks!